src/init.c

```diff
@@ -656,230 +656,151 @@ static int hyper_remove_container(char *json, int length)
-	if (*arg->data == NULL) {
-		fprintf(stderr, "allocate memory for reading file failed\n");
-		goto err;
+	if (arg->rw == WRITEFILE) {
+		fd = open(arg->file, O_CREAT| O_TRUNC| O_WRONLY, 0644);
+	} else {
+		fd = open(arg->file, O_RDONLY);
 	}
-
-	fd = open(arg->file, O_RDONLY);
 	if (fd < 0) {
 		perror("fail to open target file");
-		goto err;
-	}
-
-	fprintf(stdout, "file length %d\n", *arg->datalen);
-	while(len < *arg->datalen) {
-		size = read(fd, *arg->data + len, *arg->datalen - len);
-
-		if (size < 0) {
-			if (errno == EINTR)
-				continue;
-
-			perror("fail to read data from file");
-			goto err;
-		}
-
-		len += size;
+		goto exit;
 	}
-
-	fprintf(stdout, "read data %s\n", *arg->data);
 	ret = 0;
-err:
-	hyper_send_type(arg->pipe[1], ret ? ERROR : READY);
-	_exit(ret);
+
+exit:
+	write(arg->pipe[1], &fd, sizeof(fd));
+	exit(ret);
 }
 
-static int hyper_cmd_read_file(char *json, int length, uint32_t *datalen, uint8_t **data)
+static int hyper_cmd_rw_file(char *json, int length, uint32_t *rdatalen, uint8_t **rdata, int rw)
 {
-	struct file_command cmd;
-	struct hyper_container *c;
-	struct hyper_pod *pod = &global_pod;
+	struct file_command cmd = {
+		.id = NULL,
+		.file = NULL,
+	};
 	struct hyper_file_arg arg = {
 		.pipe = {-1, -1},
+		.rw = rw,
 	};
-	int stacksize = getpagesize() * 4;
+	struct hyper_container *c;
+	struct hyper_pod *pod = &global_pod;
+	char *data = NULL;
 	void *stack = NULL;
-	int pid, ret = -1, status;
-	uint32_t type;
+	int stacksize = getpagesize() * 4;
+	int fd = -1, len = 0, ret = -1;
+	int pid, status, datalen, size;
 
-	fprintf(stdout, "%s\n", __func__);
+	fprintf(stdout, "%s: %s\n", __func__, rw == WRITEFILE ? "write" : "read");
+
+	if (rw == WRITEFILE) {
+		// TODO: send the data via hyperstream rather than append it at the end of the command	
+		data = strchr(json, '}');
+		if (data == NULL) {
+			goto out;
+		}
+		data++;
+		datalen = length - (data - json);
+		length = data - json;
+	}
 
 	if (hyper_parse_file_command(&cmd, json, length) < 0) {
 		goto out;
 	}
+	arg.file = cmd.file;
 
 	c = hyper_find_container(pod, cmd.id);
 	if (c == NULL) {
 		fprintf(stderr, "can not find container whose id is %s\n", cmd.id);
 		goto out;
 	}
 
-	if (pipe2(arg.pipe, O_CLOEXEC) < 0) {
-		perror("create reader pipe failed");
-		goto out;
-	}
-
 	arg.mntns = c->ns;
 	if (arg.mntns < 0) {
 		perror("fail to open mnt ns");
 		goto out;
 	}
 
-	arg.file = cmd.file;
-	arg.datalen = datalen;
-	arg.data = data;
+	if (pipe2(arg.pipe, O_CLOEXEC) < 0) {
+		perror("create pipe failed");
+		goto out;
+	}
 
 	stack = malloc(stacksize);
 	if (stack == NULL) {
 		perror("fail to allocate stack for container init");
 		goto out;
 	}
 
-	pid = clone(hyper_do_cmd_read_file, stack + stacksize, CLONE_VM| SIGQUIT, &arg);
+	pid = clone(hyper_do_cmd_rw_file, stack + stacksize, CLONE_FILES | SIGCHLD, &arg);
 	if (pid < 0) {
-		perror("fail to fork reader process");
+		perror("fail to fork child process");
 		goto out;
 	}
```

Comment:
father process and child process share the same fd table, if close pipe[1] here, it won't work.